automod/rules/spammentions.go

```diff
@@ -0,0 +1,38 @@
+package rules
+
+import (
+	appbsky "github.com/bluesky-social/indigo/api/bsky"
+	"github.com/bluesky-social/indigo/automod"
+)
+
+var _ automod.PostRuleFunc = SpamMentionsRule
+
+var mentionHourlyThreshold = 20
+
+// SpamMentionsRule looks for accounts which mention an unusually large number of distinct accounts per period.
+func SpamMentionsRule(evt *automod.RecordEvent, post *appbsky.FeedPost) error {
+	did := evt.Account.Identity.DID.String()
+
+	// Increment counters for all new mentions in this post.
+	var newMentions bool
+	for _, facet := range post.Facets {
+		for _, feature := range facet.Features {
+			mention := feature.RichtextFacet_Mention
+			if mention == nil {
+				continue
+			}
+			evt.IncrementDistinct("mentions", did, mention.Did)
+			newMentions = true
+		}
+	}
+
+	// If there were any new mentions, check if it's gotten spammy.
+	if !newMentions {
+		return nil
+	}
+	if mentionHourlyThreshold <= evt.GetCountDistinct("mentions", did, automod.PeriodHour) {
```

Comment:
Interestingly, I think it's basically impossible for something to trigger on a single message, due to the order of events: counter increments get buffered until persist time, so if we're reading the counts on the same topic like this during the same rule function, we're getting the data from before we started.

That's maybe fine, but seems like maybe we should spawn a todo to write something up about that in a doc block somewhere about things to be expecting as a rule author?

Reply:
Yup! And in particular with limits you'll hit most rules on "limit plus one" not the limit number directly. Agree this should be included in rule-writing guide.